src/server/accept.rs

```diff
@@ -0,0 +1,194 @@
+use hyper_util::rt::TokioTimer;
+use std::pin::Pin;
+use std::process::ExitCode;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::net::TcpListener;
+use tokio::sync::{watch, Semaphore};
+use tokio::task::JoinSet;
+use tokio::{signal, time};
+
+#[cfg(feature = "http2")]
+use hyper_util::rt::TokioExecutor;
+
+use super::acceptor::Acceptor;
+use super::stream::IoStream;
+use crate::app::{App, AppService};
+use crate::error::ServerError;
+
+macro_rules! joined {
+    ($result:expr) => {
+        match $result {
+            // Succussfully joined the connection.
+            Ok(Ok(_)) => {}
+            // The connection was cancelled or the panicked.
+            Err(error) => {
+                if error.is_panic() {
+                    // Placeholder for tracing...
+                    if cfg!(debug_assertions) {
+                        eprintln!("error(connection): {}", error);
+                    }
+                }
+            }
+            // An error occurred that originates from hyper or tokio.
+            Ok(Err(error)) => {
+                // Placeholder for tracing...
+                if cfg!(debug_assertions) {
+                    eprintln!("error(connection): {}", error);
+                }
+            }
+        }
+    };
+}
+
+pub async fn accept<A, T>(
+    listener: TcpListener,
+    acceptor: A,
+    app: App<T>,
+    max_body_size: usize,
+    max_connections: usize,
+    shutdown_timeout: u64,
+) -> ExitCode
+where
+    A: Acceptor + Send + Sync + 'static,
+    T: Send + Sync + 'static,
+{
+    // Create a semaphore with a number of permits equal to the maximum number
+    // of connections that the server can handle concurrently. If the maximum
+    // number of connections is reached, we'll wait until a permit is available
+    // before accepting a new connection.
+    let semaphore = Arc::new(Semaphore::new(max_connections));
+
+    // Create a watch channel to notify the connections to initiate a
+    // graceful shutdown process when the `ctrl_c` future resolves.
+    let mut shutdown_rx = {
+        let (tx, rx) = watch::channel(None);
+        tokio::spawn(wait_for_ctrl_c(tx));
+        rx
+    };
+
+    // Create a JoinSet to track inflight connections. We'll use this to wait for
+    // all connections to close before the server exits.
+    let mut connections = JoinSet::<Result<(), ServerError>>::new();
+
+    // Wrap app in an arc so it can be cloned into the connection task.
+    let app = Arc::new(app);
+
+    // Start accepting incoming connections.
+    let exit_code = loop {
+        // Acquire a permit from the semaphore.
+        let permit = match semaphore.clone().acquire_owned().await {
+            Ok(acquired) => acquired,
+            Err(_) => break 1.into(),
+        };
+
+        // Accept the next stream from the tcp listener.
+        let (stream, _addr) = tokio::select! {
+            biased;
+            result = time::timeout(Duration::from_secs(15), listener.accept()) => {
```

Comment:
I really like this idea and can see it being implemented in the future if the community agrees with me.

This seems a bit novel and I would like to get some feeback from everyone before exploring this approach further.